### PR TITLE
Feature add support for want to compute task inside task to compute

### PR DESCRIPTION
**Pull request by rwrzesien**
Resolves https://github.com/golemfactory/golem/issues/3272

Please take into consideration that this is my first pull request in Golem. All comments are appreciated.

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -31,7 +31,7 @@ eth-tester==0.1.0-beta.24
 eth-utils==1.0.3
 ethereum==1.6.1
 Golem-Verificator==1.4.0
-Golem-Messages==2.13.0
+Golem-Messages==2.13.0+dev4.gd8cd1d3
```
Comment (cameel) [inline requirements.txt]: I don't think you're supposed to put untagged development versions in the requirements.

File: tests/golem/network/concent/test_helpers.py [hunk 2]
```diff
@@ -19,7 +19,7 @@ def test_self_payment(self):
         ecc = ECCx(privkey)
         ecc.verify = mock.Mock()
         msg = mock.Mock()
-        msg.eth_account = encode_hex(addr)
+        msg.task_to_compute.provider_ethereum_address = encode_hex(addr)
```
Comment (cameel) [inline tests/golem/network/concent/test_helpers.py:22]: Wasn't `eth_account` supposed to be replaced by a property to avoid having this kind of changes?

File: golem/task/tasksession.py [hunk 3]
```diff
@@ -550,8 +543,21 @@ def _react_to_want_to_compute_task(self, msg):
     @history.provider_history
     def _react_to_task_to_compute(self, msg):
         ctd = msg.compute_task_def
-        if ctd is None:
-            logger.debug('TaskToCompute without ctd: %r', msg)
+        want_to_compute_task = msg.want_to_compute_task
+        if ctd is None or want_to_compute_task is None:
+            logger.debug(
+                'TaskToCompute without ctd or want_to_compute_task: %r', msg)
+            self.task_computer.session_closed()
```
Comment (cameel) [inline golem/task/tasksession.py:549]: I think that golem-messages should have a validation against this.

File: requirements.txt (hunk 1 above)
Comment (Krigpl) [reply requirements.txt]: This is fine to check whether CI passes but should be changed to a proper version before merging.

File: golem/resource/resourcehandshake.py [hunk 4]
```diff
@@ -89,6 +89,8 @@ def request_task(self, node_name, task_id, perf_index, price,
             max_memory_size=max_memory_size,
             num_cores=num_cores,
             concent_enabled=self.task_server.client.concent_service.enabled,
+            provider_public_key=key_id,
+            provider_ethereum_public_key=key_id,
```
Comment (Krigpl) [inline golem/resource/resourcehandshake.py]: Why is this needed? I have a feeling that ethereum public key shouldn't be necessary for resource handshake.

Comment (jiivan) [reply golem/resource/resourcehandshake.py]: Actually it's not really related to resources. It's just a method that generates `WantToComputeTask`. Reasoning for including provider eth details in WTC is in https://github.com/golemfactory/golem-messages/issues/244

File: requirements.txt (hunk 1 above)
Comment (shadeofblue) [reply requirements.txt]: exactly as @Krigpl said

File: golem/network/concent/helpers.py [hunk 5]
```diff
@@ -38,11 +38,13 @@ def verify_message_payment_address(
     # Prevent self payments. This check deserve its own reject_reason but also
     # it belongs ealier in the flow rather then here.
     if privtoaddr(ecc.get_privkey()) == decode_hex(
-            report_computed_task.eth_account):
+            report_computed_task.task_to_compute.provider_ethereum_address):
```
Comment (shadeofblue) [inline golem/network/concent/helpers.py]: @rwrzesien hmm, I know this had been there before - but maybe it would be good to put the `decode_hex(address)` lookup first so that it doesn't need to be converted twice (here and in line 45 below)

File: golem/resource/resourcehandshake.py (hunk 4 above)
Comment (Krigpl) [reply golem/resource/resourcehandshake.py]: Right, I got confused and thought this is some kind of Hello message instead of WantToComputetTask

File: golem/task/tasksession.py (hunk 3 above)
Comment (shadeofblue) [reply golem/task/tasksession.py:549]: adding a way to ensure this: 
https://github.com/golemfactory/golem-messages/issues/262

File: requirements.txt (hunk 1 above)
Comment (rwrzesien) [reply requirements.txt]: Yes, also my point was to allow anyone who would like to test/run from this branch to have everything what is needed in place. I didn't assume that this commit would get merged along with this branch.

File: golem/resource/resourcehandshake.py (hunk 4 above)
Comment (rwrzesien) [reply golem/resource/resourcehandshake.py]: Yes, exactly, this method accepts parameters that are passed directly to `WantToComputeTask`.

File: tests/golem/network/concent/test_helpers.py (hunk 2 above)
Comment (rwrzesien) [reply tests/golem/network/concent/test_helpers.py:22]: Yes, but please take a look on lines above, it is set on a `Mock` object to test `process_report_computed_task_no_time_check` function.

File: golem/network/concent/helpers.py (hunk 5 above)
Comment (rwrzesien) [reply golem/network/concent/helpers.py]: @shadeofblue Sure, no problem done :+1:

File: golem/task/tasksession.py [hunk 6]
```diff
@@ -550,8 +543,21 @@ def _react_to_want_to_compute_task(self, msg):
...
+        want_to_compute_task = msg.want_to_compute_task
+        if ctd is None or want_to_compute_task is None:
+            logger.debug(
+                'TaskToCompute without ctd or want_to_compute_task: %r', msg)
+            self.task_computer.session_closed()
+            self.dropped()
+            return
+
+        try:
+            msg.verify_signature(
```
Comment (shadeofblue) [inline golem/task/tasksession.py]: @rwrzesien please correct me if I'm wrong but `msg` here is a `TaskToCompute`, isn't it? so I believe it should be:

```
want_to_compute_task.verify_signature(
    decode_hex(msg.provider_public_key)  # we can use the helper here actually... 
)
```

am I missing something?

File: requirements.txt (hunk 1 above)
Comment (shadeofblue) [reply requirements.txt]: @rwrzesien btw - does it work this way? I mean - shouldn't this line be a link to github's commit? like this: https://github.com@4r/golemfactory/golem/pull/3317/files#diff-b4ef698db8ca845e5845c4618278f29aR34 ?

File: tests/golem/network/concent/test_helpers.py (hunk 2 above)
Comment (shadeofblue) [reply tests/golem/network/concent/test_helpers.py:22]: @rwrzesien @cameel (disregarding the fact that here something else is happening) hmm, well it would be better to just replace the occurences of `eth_account` with `ttc.provider_ethereum_address` anyway sooner or later... so if we can do it here, all the better...

File: requirements.txt (hunk 1 above)
Comment (shadeofblue) [reply requirements.txt]: @rwrzesien as I was afraid -> it doesn't work that way: https://buildbot.golem.network/buildbot/#/builders/9/builds/6583/steps/2/logs/install_requirements -> please replace the requirement with a link to the github commit (and once the `golem-messages` counterpart is released we'll be able to put the correct version here before merging)

File: golem/task/tasksession.py (hunk 6 above)
Comment (shadeofblue) [reply golem/task/tasksession.py]: @rwrzesien on second thought -> it may make little sense to verify the signature of the `want_to_compute_task` against the included `provider_public_key` here - remember, we're the provider and the message had been generated by us - so, it would actually make more sense to verify that the attached `WantToComputeTask` is _our_ message, so, instead, we should probably do the following:

```
want_to_compute_task.verify_signature(self.task_server.keys_auth.ecc.raw_pubkey)
```

to ensure the message had been generated by us...

File: requirements.txt (hunk 1 above)
Comment (rwrzesien) [reply requirements.txt]: :+1: done

File: golem/task/tasksession.py (hunk 6 above)
Comment (rwrzesien) [reply golem/task/tasksession.py]: :+1: done

File: requirements.txt (hunk 1 above)
Comment (rwrzesien) [reply requirements.txt]: :+1: done

File: golem/task/tasksession.py (hunk 6 above)
Comment (rwrzesien) [reply golem/task/tasksession.py]: :+1: done

File: requirements.txt (hunk 1 above)
Comment (cameel) [reply requirements.txt]: @shadeofblue Just FYI, it would work if it was what `git describe` returns (which would look like this: `2.13.0-10-gd8cd1d3`). It does not work because it's from the version script and does not have  the original `<tag>-<n>-<hash>` format. I was surprised too when I saw it for the first time but git does allow you to use something like this as a reference to a commit and simply ignores everything but the hash.

Comment (cameel) [reply requirements.txt]: Oh, and by the way - if this was just temporary change I think it would have been a good idea to have a comment in the code. Or have it in a separate commit. Otherwise it's not obvious and easy to miss later when you merge (hence my remark that started this thread).

File: golem/task/tasksession.py [hunk 7]
```diff
@@ -550,8 +542,21 @@ def _react_to_want_to_compute_task(self, msg):
     @history.provider_history
     def _react_to_task_to_compute(self, msg):
         ctd = msg.compute_task_def
-        if ctd is None:
-            logger.debug('TaskToCompute without ctd: %r', msg)
+        want_to_compute_task = msg.want_to_compute_task
+        if ctd is None or want_to_compute_task is None:
```
Comment (shadeofblue) [inline golem/task/tasksession.py:546]: @rwrzesien so yeah, I believe the relevant validation _has_ been added to `golem-messages`

File: tests/golem/network/concent/test_received_handler.py [hunk 8]
```diff
@@ -238,11 +238,12 @@ def get_vrct(self):
             requestor_public_key=msg_utils.encode_hex(
                 self.requestor_keys.raw_pubkey,
             ),
-            provider_public_key=msg_utils.encode_hex(
-                self.provider_keys.raw_pubkey,
-            ),
             sign__privkey=self.requestor_keys.raw_privkey,
         )
+        ttc.want_to_compute_task.provider_public_key = msg_utils.encode_hex(
```
Comment (shadeofblue) [inline tests/golem/network/concent/test_received_handler.py]: @rwrzesien I believe it would be safer to specify the relevant key in the call to `TaskToComputeFactory` above with:

```
want_to_compute_task__provider_ethereum_public_key = self.provider_keys.raw_pubkey,
want_to_compute_task__sign__privkey=self.provider_keys.raw_privkey
```

to ensure the `WantToComputeTask` is signed correctly... but maybe the handlers here don't check that condition...

**Review by shadeofblue — APPROVED** [review]
seems good - I added one comment above to make the test more futureproof...

File: golem/task/tasksession.py (hunk 7 above)
Comment (rwrzesien) [reply golem/task/tasksession.py:546]: @shadeofblue Should I remove this check then? If yes, should I remove it along with check for CTD which was there before?

File: tests/golem/network/concent/test_received_handler.py (hunk 8 above)
Comment (rwrzesien) [reply tests/golem/network/concent/test_received_handler.py]: :+1: done

**Review by jiivan — APPROVED** [review]
LGTM
